### PR TITLE
feat: report status of java compilation as `stage/Build`

### PR DESCRIPTION
**Pull request by keturn**
as a replacement for CloudBees reporting every Stage.

This is currently on Nanoware/develop. Example: https://github.com/Nanoware/Health/pull/1

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -69,11 +73,20 @@ pipeline {
         stage('Build') {
             steps {
                 // Jenkins sometimes doesn't run Gradle automatically in plain console mode, so make it explicit
-                sh './gradlew --console=plain clean htmlDependencyReport jar'
+                sh label: 'determining dependencies', script: './gradlew --console=plain clean htmlDependencyReport'
+
+                // Reporting this compile task as `stage/Build` for backwards compatibility with things configured
+                // to watch for CloudBees SCM Reporting stage status.
+                gitStatusWrapper(gitHubContext: 'stage/Build', description: 'Compiling Java') {
+                    sh './gradlew --console=plain jar'
```
Comment (keturn) [inline Jenkinsfile:81]: This `gitStatusWrapper` is the main thing here, that's what will provide a `stage/Build` status to match against.

File: Jenkinsfile [hunk 2]
```diff
@@ -69,11 +73,20 @@ pipeline {
...
+                }
                 archiveArtifacts 'build/libs/*.jar'
             }
             post {
                 always {
+                    recordIssues enabledForFailure: true, failOnError: true, publishAllIssues: true, skipBlames: true,
+                        qualityGates: [[threshold: 1, type: 'TOTAL_ERROR', unstable: false]],
+                        tools: [java()]
```
Comment (keturn) [inline Jenkinsfile:89]: The `recordIssues java` isn't required, but it does get the compilation errors to show up as annotation checks.

File: Jenkinsfile [hunk 3]
```diff
@@ -49,6 +49,10 @@ pipeline {
         stage('Setup') {
             steps {
                 echo 'Automatically checked out the things!'
+                // the default resolution when omitting `defaultBranch` is to `master`
+                // this is wrong in our case, so explicitly set `develop` as default
+                // TODO: does this also work for PRs with different base branch?
+                discoverGitReferenceBuild(defaultBranch: 'develop')
```
Comment (jdrueckert) [inline Jenkinsfile:55]: Does doing it in the "Setup" stage make it usable throughout the whole pipeline? If so: yey and totally makes sense to put it here!

File: Jenkinsfile (hunk 1 above)
Comment (jdrueckert) [reply Jenkinsfile:81]: as I mentioned yesterday, I'd also be fine with not having the `stage/*` reports as long as we have an automated way to align the required checks in all repos. But as long as we don't I think it's nice to have this, thanks for looking into it 👍

File: Jenkinsfile (hunk 2 above)
Comment (jdrueckert) [reply Jenkinsfile:89]: which is very helpful indeed 👍

File: Jenkinsfile (hunk 3 above)
Comment (keturn) [reply Jenkinsfile:55]: I think that should apply to everything after it's run, yes. There isn't much isolation at all between the Stages.

**Review by jdrueckert — APPROVED** [review]

**Comment by keturn** [comment]
[Jenkins results as GitHub checks (GitHub App Authentication)](https://trello.com/c/t1FojXzr/126-jenkins-results-as-github-checks-github-app-authentication)

**Pull request merged**